### PR TITLE
Restrict Unauthorized Users from Accessing Facility's Patient Registration

### LINKED ISSUE #7244: Prevent users who don't have access to register a patient to a facility from opening the patient registration page of that facility
**Describe the bug**
Certain user types are not allowed to register a patient into a non-home facility. However, it is frustrating when user fills out the patient form only to realise when submitting that they don't have permission to do this action.

**To Reproduce**

1. Register a patient to a non home facility as a doctor
2. Fill the patient registration form and hit submit
3. See permission denied error

**Expected behavior**
Users who do not have permission to register a patient to a non-home facility should not be allowed to open patient registration form for that facility.

**Screenshots**
image

### PR DESCRIPTION
**Pull request by AshrafMd-1**
## Proposed Changes

- Fixes #7244 

@coronasafe/care-fe-code-reviewers @coronasafe/code-reviewers

## Merge Checklist

- [ ] Add specs that demonstrate bug / test a new feature.
- [ ] Update [product documentation](https://docs.coronasafe.network/coronasafe-care-documentation/architecture/architecture-and-layering-of-care).
- [ ] Ensure that UI text is kept in I18n files.
- [ ] Prep screenshot or demo video for changelog entry, and attach it to issue.
- [ ] Request for Peer Reviews
- [ ] Completion of QA

### CODE REVIEW

**Comment by AshrafMd-1** [comment]
@rithviknishad can you check once

**Comment by AshrafMd-1** [comment]
@rithviknishad can you remove the merge conflict

File: src/Components/Patient/PatientRegister.tsx [hunk 1]
```diff
@@ -1060,6 +1061,32 @@ export const PatientRegister = (props: PatientRegisterProps) => {
...
+      authUser.state === facilityObject?.state
+    ) {
+      check = true;
+    }
+    return check;
+  };
+
+  if (!isLoading && facilityId && facilityObject && !PatientRegisterAuth()) {
+    return <Error404 />;
+  }
```
Comment (rithviknishad) [inline src/Components/Patient/PatientRegister.tsx]: This looks like an unnecessary complication...
We should not be redirected to this page in the first place if a user is not allowed to.
```suggestion
```

**Review by rithviknishad — CHANGES_REQUESTED** [review]

File: src/Components/Patient/PatientRegister.tsx (hunk 1 above)
Comment (AshrafMd-1) [reply src/Components/Patient/PatientRegister.tsx]: @rithviknishad Normally you will not be redirected without access but suppose you paste a URL of another facility then this code Is important

**Review by rithviknishad — APPROVED** [review]

**Comment by nihal467** [comment]
@AshrafMd-1 fix the lint issue

**Review by Ashesh3 — APPROVED** [review]

**Comment by AshrafMd-1** [comment]
@rithviknishad can you change the labels

**Comment by nihal467** [comment]
test set for restriction

- [x] district admin state admin and home nurse are able to create a patient
- [x] linked doctors and non-home nurse are not able to create a patient 
- [x] verify linked doctors are able to view patient datas

**Comment by nihal467** [comment]
image

image

image


- [ ] When I go to a patient details page under my linked facility and click on the patient detail button, I am redirected to a "page not found" page.
- [ ] Similarly, if I go to the patient advanced filters, apply the facility filter for my linked facility, then click the "add patient" button, I am redirected to a "page not found" page.

**Expected behavior**

@AshrafMd-1 when they try to access them, keep them on the same page, and show the error " Oops! Non-Home facility users don't have permission to perform this action."

CC: @rithviknishad @Ashesh3

**Comment by AshrafMd-1** [comment]
@nihal467 @rithviknishad  can you check once if it is okay or are there any shortcomings?

File: src/Components/ExternalResult/FacilitiesSelectDialogue.tsx [hunk 2]
```diff
@@ -29,6 +31,16 @@ const FacilitiesSelectDialog = (props: Props) => {
...
+        district={
+          authUser?.user_type === "DistrictAdmin"
+            ? authUser?.district?.toString()
+            : ""
+        }
+        state={
+          authUser?.user_type === "StateAdmin"
+            ? authUser?.state?.toString()
+            : ""
+        }
```
Comment (rithviknishad) [inline src/Components/ExternalResult/FacilitiesSelectDialogue.tsx]: pass `null/undefined` instead of empty string?

File: src/Components/Patient/PatientRegister.tsx [hunk 3]
```diff
@@ -1061,6 +1062,32 @@ export const PatientRegister = (props: PatientRegisterProps) => {
...
+      authUser.district === facilityObject?.district
+    ) {
+      check = true;
+    } else if (
+      authUser.user_type === "StateAdmin" &&
+      authUser.state === facilityObject?.state
+    ) {
+      check = true;
+    }
+    return check;
```
Comment (rithviknishad) [inline src/Components/Patient/PatientRegister.tsx]:
```suggestion
    const showAllFacilityUsers = ["DistrictAdmin", "StateAdmin"];
    if (
      !showAllFacilityUsers.includes(authUser.user_type) &&
      authUser.home_facility_object?.id === facilityId
    ) {
      return true;
    }
    
    if (
      authUser.user_type === "DistrictAdmin" &&
      authUser.district === facilityObject?.district
    ) {
      return true;
    }
    
    if (
      authUser.user_type === "StateAdmin" &&
      authUser.state === facilityObject?.state
    ) {
      return true
    }
    
    return false;
```

**Review by rithviknishad — CHANGES_REQUESTED** [review]

**Comment by AshrafMd-1** [comment]
@rithviknishad can you check once

**Review by rithviknishad — APPROVED** [review]

**Comment by nihal467** [comment]
@AshrafMd-1 
image

- [ ] When we click on the add patient button with a nurse user having 1 home facility and multiple linked facility, the error is thrown.

Note: Before marking it again for testing, do a QA test for self-evaluation of the PR

CC: @AshrafMd-1

**Comment by AshrafMd-1** [comment]
@nihal467 my bad i did check all those edge cases but forgot to test main case 

My exams will be over by tomorrow so by the end of Saturday I will submit a tested pr

**Comment by gigincg** [comment]
@AshrafMd-1 Can you update this PR?

**Comment by AshrafMd-1** [comment]
@nihal467 can you check once 

cc: @rithviknishad

**Comment by AshrafMd-1** [comment]
ready for testing

**Comment by nihal467** [comment]
LGTM

**Review by khavinshankar — APPROVED** [review]
